### ISSUE TITLE
Markup is partially lost

### ISSUE THREAD

**Issue by thany** [vector-im/element-web#17242]
### Description

Some markup is lost during sending a message, and other markup may be lost only after editing.

### Steps to reproduce

1. Type this message, EXACTLY:

```
Hi, I am line 1
I am line 2
And \*here\* is line 3
```

2. Send it
3. Observe the newlines being lost.
4. Edit it. Apparently it's neccesary to add two spaces ad the end of each line to make actual newlines. Please do so.
5. Send it.
6. Observe the literal asterisks being made un-literal, so `\*here\*` becomes *`here`*
7. Edit it
8. Observe the previously typed asterisks are now inexplicably underscores.
9. AND the two-space newlines have been destroyed.

Logs being sent: no (I don't think they're needed, since this is a generic issue - nothing to do with me or my setup)

### Version information

- **Platform**: desktop

For the desktop app:

- **OS**: Windows 10 20H2
- **Version**: 1.7.27

**Comment by t3chguy** [comment]
Converting Commonmark -> HTML -> Commonmark will always be lossy.

**Comment by thany** [comment]
> 
> 
> Converting Commonmark -> HTML -> Commonmark will always be lossy.

Well then, don't?

**Comment by t3chguy** [comment]
Element accepts Commonmark, Matrix only plaintext or a Matrix-defined HTML subset.
So if you'd have to force users to input/edit HTML for no conversions to take place.

**Comment by thany** [comment]
Well, plaintext it is then.

**Comment by t3chguy** [comment]
Sure, you're more than able to send plaintext only messages now and then you'll have no issue whilst editing.
You can use `/plain` to ensure your messages won't be parsed as Markdown (Commonmark)

**Comment by thany** [comment]
No no, they do need to be markdown, as displayed as such. Just don't do markdown -> html -> markdown.

**Comment by t3chguy** [comment]
The matrix `body` field is defined as plaintext, the output of one Markdown dialect is likely different in another Markdown dialect, hence the markdown conversion on send rather than on receipt.

**Comment by thany** [comment]
In a standardized system you shouldn't have to assume implementation of such standard is different across clients. If Matrix defines one dialect of markdown, then that dialect must be used, and we must assume any other client uses the same dialect.

With this in mind, even if markdown-html-markdown conversion is deemed neccesary, it shouldn't have to lead to data loss.

Having said that, assuming markdown dialect is standardized, what does it matter if the `body` is populated with literal markdown? As long as it is parsed as such, there's no problem. If it's html, then parse it as HTML, but don't try to convert to markdown, because that causes data loss.

To recap:

* Type markdown -> send as plaintext -> parse as markdown
* Type plaintext -> send as plaintext -> parse as markdown (because invalid markdown is displayed as-is)
* Someone capable of sending HTML -> parse as HTML -> edit as HTML -> send as HTML

Since Element is not able to compose HTML messages, I doubt that even *considering* HTML is actually useful. If everyone uses markdown, why are we even discussing HTML? Surely, no sensible client would ever let the user compose HTML...

**Comment by t3chguy** [comment]
Matrix doesn't support markdown over the wire, so the standard can only go as far as Element.
If you wish to suggest Matrix have a standard Markdown dialect then the repo to suggest that would be https://github.com/matrix-org/matrix-doc/ and not here.

> Since Element is not able to compose HTML messages

Sure it can, you can enter tags like `<del>`, `<u>` etc which are not part of Commonmark but are valid in Element
Commands like `/rainbow` also output HTML which is impossible to express in any common markdown dialect.
There is of course also the `/html` command to send your input verbatim.

**Comment by thany** [comment]
Matrix supports plaintext, and plaintext is markdown is plaintext is markdown. The two are 100% interchangable.
The same can not be said about HTML.

So, to repeat myself again: Don't ever convert to HTML. Period. Always always always keep messages in plaintext and interpret them as markdown. Just don't do anything supposedly intelligent in the client. How hard can it be?

**Comment by t3chguy** [comment]
That would mean someone trying to send a plaintext message, from a client that doesn't render formatting, e.g a terminal client or IRC bridge would now have to consider what that message might look like to the end user, if a user wanted to write a `+` at the start of a line it'd get turned into a bullet point, etc.

**Comment by piegamesde** [comment]
> Converting Commonmark -> HTML -> Commonmark will always be lossy.

It doesn't have to. HTML is a super set of Commonmark feature wise, but the HTML used for editing is already restricted to what Commonmark supports due to the first conversion. A proper conversion implementation would do the correct escaping so that there is no loss.

However, I acknowledge that the conversion will always be *transformative*, as in stars may become underscores to make italics etc. This not a big deal.

**Comment by t3chguy** [comment]
Also the fact that Element in Markdown mode doesn't let you do freeform HTML like tables, those can only be done using `/html` - so when you edit you would need to be shown `/html <bunch of HTML>` instead of markdown whenever it encounters something which qualifies, so still would be lossy

**Comment by piegamesde** [comment]
> so still would be lossy

*Fine.* But there are cases, where it is lossy, but it absolutely doesn't need to, and this is a bug, and please make it *less* lossy. There are big, low-hanging fruits here and we can talk about the edge cases later.

**Comment by t3chguy** [comment]
Duplicate of https://github.com/vector-im/element-web/issues/10725
